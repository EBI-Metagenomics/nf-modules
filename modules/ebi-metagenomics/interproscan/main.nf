process INTERPROSCAN {
    tag "$meta.id"
    label 'process_long'

    container 'microbiome-informatics/interproscan:5.73-104.0'

    containerOptions {
        def containerArgs = []
        def mountArg = (workflow.containerEngine == 'singularity') ? "--bind" : "--volume"

        containerArgs << "${mountArg} ${task.workDir}/${interproscan_db}/data:/opt/interproscan/data"

        if ( params.interpro_licensed_software ) {
            def licensedSoftwarePath = "${task.workDir}/${interproscan_db}/licensed"
            containerArgs << "${mountArg} ${licensedSoftwarePath}:/opt/interproscan/licensed"
            // This override is needed otherwise it fails because this path seems to be hardcoded in the container
            containerArgs << "${mountArg} ${licensedSoftwarePath}/signalp:/usr/opt/www/pub/CBS/services/SignalP-4.1/signalp-4.1"
        }

        return containerArgs.join(' ')
    }

    input:
    tuple val(meta), path(fasta)
    tuple path(interproscan_db), val(db_version)

    output:
    tuple val(meta), path('*.tsv.gz') , emit: tsv
    tuple val(meta), path('*.xml.gz') , emit: xml
    tuple val(meta), path('*.gff3.gz'), emit: gff3
    tuple val(meta), path('*.json.gz'), emit: json
    path "versions.yml"               , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"
    def is_compressed = fasta.extension == "gz"
    def fasta_file_name = fasta.name - ~/\.gz$/
    """
    if [ "$is_compressed" == "true" ]; then
        gzip -c -d ${fasta} > ${fasta_file_name}
    fi

    # -dp (disable precalculation) is on so no online dependency
    interproscan.sh \\
        -cpu $task.cpus \\
        -i ${fasta_file_name} \\
        -dp \\
        ${args} \\
        --output-file-base ${prefix}

    gzip ${prefix}.{tsv,xml,gff3,json}

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        InterProScan: \$(interproscan.sh --version | grep -o "InterProScan version [0-9.-]*" | sed "s/InterProScan version //")
        InterProScan database: $db_version
    END_VERSIONS
    """

    stub:
    def prefix = task.ext.prefix ?: "${meta.id}"
    """
    echo '' > ${prefix}.{tsv,xml,gff3,json}
    gzip ${prefix}.{tsv,xml,gff3,json}

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        InterProScan: \$(interproscan.sh --version | grep -o "InterProScan version [0-9.-]*" | sed "s/InterProScan version //")
        InterProScan database: $db_version
    END_VERSIONS
    """
}
