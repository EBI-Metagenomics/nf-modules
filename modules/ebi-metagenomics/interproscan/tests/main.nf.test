
nextflow_process {

    name "Test Process INTERPROSCAN"
    script "../main.nf"
    process "INTERPROSCAN"

    tag "modules"
    tag "modules_ebimetagenomics"
    tag "interproscan"

    test("test_interproscan") {

        when {
            process {
                """
                input[0] = [
                    [ id:'test', single_end:false ], // meta map
                    file("${moduleDir}/tests/fixtures/test.fasta.gz", checkIfExists: true)
                ]
                input[1] = [
                    file("${moduleDir}/tests/fixtures/interproscan_db/", type: 'dir', checkIfExists: true),
                    "5.73-104.0"
                ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }
}
