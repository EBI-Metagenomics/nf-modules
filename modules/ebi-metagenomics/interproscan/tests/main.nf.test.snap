{
    "test_interproscan": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.tsv.gz:md5,1a3983f24f0018d9c1515fe91a1a7237"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.xml.gz:md5,85b6e298f4853070ee4014301c07150f"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.gff3.gz:md5,964c009c6a5b22edecc336efcd6ca5bd"
                    ]
                ],
                "3": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.json.gz:md5,8c91c66a974f22b5bff02019b6529229"
                    ]
                ],
                "4": [
                    "versions.yml:md5,7f61cf700565d5b1ab2ca54b0b72fc71"
                ],
                "gff3": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.gff3.gz:md5,964c009c6a5b22edecc336efcd6ca5bd"
                    ]
                ],
                "json": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.json.gz:md5,8c91c66a974f22b5bff02019b6529229"
                    ]
                ],
                "tsv": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.tsv.gz:md5,1a3983f24f0018d9c1515fe91a1a7237"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,7f61cf700565d5b1ab2ca54b0b72fc71"
                ],
                "xml": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.xml.gz:md5,85b6e298f4853070ee4014301c07150f"
                    ]
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.4"
        },
        "timestamp": "2025-03-04T21:52:55.725908"
    }
}